enables module modification during repackaging

PaliC opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71520

It is possible for someone to export a package, change something in it, and then attempt to repackage it. In this case it's possible that dependencies of the package that were interned before are no longer interned. In this case it is not obvious where we would look for these packages, therefore, we throw an error.

Differential Revision: [D33675557](https://our.internmc.facebook.com/intern/diff/D33675557)

PaliC committed:
enables module modification during repackaging

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33675557).

suo commented:
Mind adding some more info to the description? e.g. what was the bug, what is the fix. Would help me understand how to review

PaliC committed:
Update on "enables module modification during repackaging"


Differential Revision: [D33675557](https://our.internmc.facebook.com/intern/diff/D33675557)

[ghstack-poisoned]

bradleyhd commented on torch/package/package_importer.py:
```diff
@@ -406,14 +406,21 @@ def _do_find_and_load(self, name):
         return module
 
+    def _is_interned(self, module):
+        spec = getattr(module, 'spec', None)
```
`__spec__`

bradleyhd commented on torch/package/package_importer.py:
```diff
@@ -425,6 +432,10 @@ def _find_and_load(self, name):
             self.modules["typing.re"] = cast(Any, module).re
 
+        if filename is None and self._is_interned(module):
+            message = f"import of {name} halted as it's source is was not interned"
```
super-nit: 'its source'

bradleyhd reviewed (changes requested):
LGTM, confirmed works against my test case in D33662192 with the one minor change

PaliC committed:
Update on "enables module modification during repackaging"


It is possible for someone to export a package, change something in it, and then attempt to repackage it. In this case it's possible that dependencies of the package that were interned before are no longer interned. In this case it is not obvious where we would look for these packages, therefore, we throw an error.

Differential Revision: [D33675557](https://our.internmc.facebook.com/intern/diff/D33675557)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33675557).

bradleyhd reviewed (changes requested):
ran into a small issue. also mind rebasing for internal tests?

PaliC commented on torch/package/package_importer.py:
```diff
@@ -406,14 +406,21 @@ def _do_find_and_load(self, name):
+        if spec is None:
+            return False
+        origin = getattr(spec, 'origin', None)
+        return origin == "<package_importer>" and not hasattr(module, "__path__")
```
Yeah, I was just going for consistency with PackageExporter. Also do you know if this is well-formed python. In the docs a package was effectively defined by the existence of a __path__ variable and an iterable in it. I'm not sure if something has to be in the iterable?

bradleyhd replied:
https://docs.python.org/3/reference/import.html#package-path-rules
> __path__ must be an iterable of strings, but it may be empty.

bradleyhd replied:
I think we're fine 👍

suo commented:
Okay, I'm catching up on this. So is the issue that the "empty" package node from package1 is getting picked up by OrderedImporter, and thus failing because really we want to use the module from package2?

If so, I think this change is probably not what we want—it should still be valid to import the empty package node in other cases. If we raise a ModuleNotFoundError when encounter an empty package node, does that not break every import of like `a.b` where we didn't actually intern `a`? I'm surprised that tests are not breaking with this change.

I think the fix should be located in how OrderedImporter resolves modules, which seems to be what D33662192 does. I am not a huge fan of the "dummy_module" thing but I guess it's fine?

PaliC committed:
Update on "enables module modification during repackaging"


It is possible for someone to export a package, change something in it, and then attempt to repackage it. In this case it's possible that dependencies of the package that were interned before are no longer interned. In this case it is not obvious where we would look for these packages, therefore, we throw an error.

Differential Revision: [D33675557](https://our.internmc.facebook.com/intern/diff/D33675557)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33675557).

suo reviewed (changes requested):
back to you per my comment, feel free to re-request

bradleyhd commented on torch/package/importer.py:
```diff
@@ -184,6 +184,13 @@ class OrderedImporter(Importer):
         self._importers: List[Importer] = list(args)
 
+    def _check_if_malformed_package(self, module):
+        if not hasattr(module, '__path__'):
```
https://docs.python.org/3/reference/import.html#path__ mentions `__path__` has to be set for packages, so should this return True? or is it that it returns False because without `__path__` the module is not a package at all, rather just a module?

PaliC replied:
ohh shoot just bad naming on my end. Maybe malformed module would be better?

PaliC replied:
Well if it's not a package, it definitely can't be a malformed package which was my logic here haha. But yeah we do need a better name, I'll just go with fileless_package for now

PaliC committed:
Update on "enables module modification during repackaging"


It is possible for someone to export a package, change something in it, and then attempt to repackage it. In this case it's possible that dependencies of the package that were interned before are no longer interned. In this case it is not obvious where we would look for these packages, therefore, we throw an error.

Differential Revision: [D33675557](https://our.internmc.facebook.com/intern/diff/D33675557)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33675557).
